### PR TITLE
Base datetime cursor state off latest observed record

### PR DESCRIPTION
**Pull request by ambirdsall**
## What
Writing the code before the description, so watch this space for updates; part 1/3 for closing https://github.com/airbytehq/airbyte-internal-issues/issues/6314

- Part 1 (this PR): add the new `Cursor.observe` method and update `DatetimeBasedCursor` to use it for more less lossy state messages; preserve the old interface to limit the blast radius of required changes to just `DatetimeBasedCursor` and `PerPartitionCursor` (which needs to correctly delegate `per_slice_cursor.observe` calls inside `PerPartitionCursor.observe`). This appears to be done and working already.
- Part 2: Update the `Cursor.close_slice` method to force all `Cursor` subclasses to use `observe` for state tracking. These changes could perhaps be merged into this PR instead.
- Part 3: Update the greenhouse and chargebee source's custom cursor implementations to use the new interface. This cannot be merged with earlier PRs since the already-published CDK version is used to validate connector changes.

## How
*Describe the solution*

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
*Are there any breaking changes? What is the end result perceived by the user?*

*For connector PRs, use this section to explain which type of semantic versioning bump occurs as a result of the changes. Refer to our [Semantic Versioning for Connectors](https://docs.airbyte.com/contributing-to-airbyte/#semantic-versioning-for-connectors) guidelines for more information. **Breaking changes to connectors must be documented by an Airbyte engineer (PR author, or reviewer for community PRs) by using the [Breaking Change Release Playbook](https://docs.google.com/document/d/1VYQggHbL_PN0dDDu7rCyzBLGRtX-R3cpwXaY8QxEgzw/edit).***

*If there are breaking changes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.*

## Pre-merge Actions
*Expand the relevant checklist and delete the others.*

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- Connector version is set to `0.0.1`
    - `Dockerfile` has version `0.0.1`
- Documentation updated
    - Connector's `README.md`
    - Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - `docs/integrations/<source or destination>/<name>.md` including changelog with an entry for the initial version. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - `docs/integrations/README.md`

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Connector Generator</strong></summary>

- Issue acceptance criteria met
- PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/resources/pull-requests-handbook)
- If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:generateScaffolds` then checking in your changes
- Documentation which references the generator is updated as needed

</details>

<details><summary><strong>Updating the Python CDK</strong></summary>

### Airbyter

Before merging:
- Pull Request description explains what problem it is solving
- Code change is unit tested
- Build and my-py check pass
- Smoke test the change on at least one affected connector
   - On Github: Run [this workflow](https://github.com/airbytehq/airbyte/actions/workflows/connectors_tests.yml), passing `--use-local-cdk --name=source-<connector>` as options
   - Locally: `airbyte-ci connectors --use-local-cdk --name=source-<connector> test`
- PR is reviewed and approved
      
After merging:
- [Publish the CDK](https://github.com/airbytehq/airbyte/actions/workflows/publish-cdk-command-manually.yml)
   - The CDK does not follow proper semantic versioning. Choose minor if this the change has significant user impact or is a breaking change. Choose patch otherwise.
   - Write a thoughtful changelog message so we know what was updated.
- Merge the platform PR that was auto-created for updating the Connector Builder's CDK version
   - This step is optional if the change does not affect the connector builder or declarative connectors.

</details>

### CODE REVIEW

**Review by girarda — APPROVED** [review]
looks great! Approved pending fixing the mypy type checks and format errors.

You can run `bin/run-mypy-on-modified-files.sh` from `airbyte-cdk/python` to run mypy  locally.

You can run ` airbyte-ci format fix python` from the root of the repository to fix the format

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py [hunk 1]
```diff
@@ -109,20 +114,39 @@ def set_initial_state(self, stream_state: StreamState) -> None:
...
+          stream state may need to be deferred depending on whether the source reliably orders records by the cursor field.
+        """
+        record_cursor_value = record.get(self._cursor_field.eval(self.config))
+        # if the current record has no cursor value, we cannot meaningfully update the state based on it, so there is nothing more to do
+        if not record_cursor_value:
+            return
+
+        start_field = self._partition_field_start.eval(self.config)
+        end_field = self._partition_field_end.eval(self.config)
+        is_highest_observed_cursor_value = not self._highest_observed_cursor_field_value or self.parse_date(
```
Comment (bleonard) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py:132]: The issue https://github.com/airbytehq/airbyte-internal-issues/issues/6314 talks about "We should instead always use the cursor value of the most recent record, if the record is within the current slice."

I don't see anything in here about "if the record is within the current slice" - I apologize if I am missing it.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py [hunk 2]
```diff
@@ -109,20 +114,39 @@ def set_initial_state(self, stream_state: StreamState) -> None:
...
-        potential_cursor_values = [
-            cursor_value for cursor_value in [self._cursor, last_record_cursor_value, stream_slice_value_end] if cursor_value
-        ]
         cursor_value_str_by_cursor_value_datetime = dict(
             map(
                 # we need to ensure the cursor value is preserved as is in the state else the CATs might complain of something like
                 # 2023-01-04T17:30:19.000Z' <= '2023-01-04T17:30:19.000000Z'
                 lambda datetime_str: (self.parse_date(datetime_str), datetime_str),
-                potential_cursor_values,
+                filter(lambda item: item, [self._cursor, self._highest_observed_cursor_field_value]),
```
Comment (bleonard) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py:149]: The issue https://github.com/airbytehq/airbyte-internal-issues/issues/6314 talks about "We should instead always use the cursor value of the most recent record, if the record is within the current slice."

I don't see anything in here about "if the record is within the current slice" - I apologize if I am missing it.

Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py:149]: this is done by checking if the cursor value is within the slice boundaries https://github.com/airbytehq/airbyte/pull/35843/files#diff-51246ab76c66444ad999d4d0e0b1590c6b1814f3d1b27b9d9441b999583a6e0dR136

Comment (bleonard) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py:149]: Unless "current" means “the one being processed” or “within the current time in the world” - Maybe words are just hard. Carry on.
Example: so at the end of the Jan 2021 slice the state could be Jan 28th if that happens to be the last record.

Comment (ambirdsall) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py:149]: > Example: so at the end of the Jan 2021 slice the state could be Jan 28th if that happens to be the last record.

@bleonard exactly so

**Pull request merged**